// @portis/web3/packages/portis-eos/lib/networks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function networkAdapter(network, gasRelay) {
    var networkObj = typeof network === 'string' ? Object.assign({}, networks[network]) : network;
    if (typeof networkObj !== 'object') {
        throw new Error("[Portis] illegal 'network' parameter. Read more about it here: https://docs.portis.io/#/configuration?id=network");
    }
    if (!networkObj.nodeUrl) {
        throw new Error("[Portis] 'nodeUrl' is required. Read more about it here: https://docs.portis.io/#/configuration?id=network");
    }
    if (gasRelay && !networkObj.gasRelayHubAddress) {
        throw new Error("[Portis] can't find default gas relay hub for " + network);
    }
    if (typeof network === 'string' && !gasRelay) {
        delete networkObj.gasRelayHubAddress;
    }
    networkObj.nodeProtocol = networkObj.nodeProtocol || 'rpc';
    return networkObj;
}
exports.networkAdapter = networkAdapter;
var networks = {
    eos: {
        nodeUrl: 'https://api.eosdetroit.io',
        nodeProtocol: 'rpc',
        chainId: 'aca376f206b8fc25a6ed44dbdc66547c36c6c33e3a119ffbeaef943642f0e906',
    },
    eosJungle: {
        nodeUrl: 'https://jungle2.cryptolions.io',
        nodeProtocol: 'rpc',
        chainId: 'e70aaab8997e1dfce58fbfac80cbbb8fecec7b99cf982a9444273cbc64c41473',
    },
};
